src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+            let typed_function = TypedNativeFunction::type_native_function(native_function);
+            Some(typed_function.type_check_appliction(self, args, context))
+        } else {
+            None
+        }
+    }
+
+    fn type_check_function_application(&mut self, expression: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+        if let Some((function_name, args)) = expression.split_first() {
+            self.type_map.set_type(function_name, no_type())?;
+            let function_name = function_name.match_atom()
+                .ok_or(CheckError::new(CheckErrors::NonFunctionApplication))?;
+
+            if let Some(type_result) = self.try_native_function_check(function_name, args, context) {
+                type_result
+            } else {
+                let function_type = self.get_function_type(function_name)
+                    .ok_or(CheckError::new(CheckErrors::UnknownFunction(function_name.clone())))?;
+                self.type_check_function_type(&function_type, args, context)
+            }
+        } else {
+            Err(CheckError::new(CheckErrors::NonFunctionApplication))
+        }
+    }
+
+    pub fn type_check(&mut self, expr: &SymbolicExpression, context: &TypingContext) -> TypeResult {
+        let mut result = self.inner_type_check(expr, context);
+
+        if let Err(ref mut error) = result {
+            if !error.has_expression() {
+                error.set_expression(expr);
+            }
+        }
+
+        result
+    }
+
+    fn lookup_variable(&self, name: &str, context: &TypingContext) -> TypeResult {
+        if let Some(type_result) = type_reserved_variable(name) {
+            Ok(type_result)
+        } else if let Some(type_result) = self.contract_context.get_variable_type(name) {
+            Ok(type_result.clone())
+        } else if let Some(type_result) = context.lookup_variable_type(name) {
+            Ok(type_result.clone())
+        } else {
+            Err(CheckError::new(CheckErrors::UnboundVariable(name.to_string())))
+        }
+    }
+
+    fn inner_type_check(&mut self, expr: &SymbolicExpression, context: &TypingContext) -> TypeResult {
+        let type_sig = match expr.expr {
+            AtomValue(ref value) => {
+                TypeSignature::type_of(value)
+            },
+            Atom(ref name) => {
+                self.lookup_variable(name, context)?
+            },
+            List(ref expression) => {
+                self.type_check_function_application(expression, context)?
+            }
+        };
+
+        self.type_map.set_type(expr, type_sig.clone())?;
+        Ok(type_sig)
+    }
+
+    fn type_check_define_variable(&mut self, args: &[SymbolicExpression], context: &mut TypingContext) -> CheckResult<(String, TypeSignature)> {
+        if args.len() != 2 {
+            return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(2, args.len())))
```

Comment:
To improve error message clarity for the programmer, would it make sense to treat this as a separate class of `IncorrectArgumentCount` error, that is specific to variables?

Reply:
Probably -- I think a lot of this could be achieved with a better Display handler for the IncorrectArgumentCount error type -- namely, just printing a message like

```
(define ...) used to define variable expects two arguments. 
```

And perhaps even using messages from the docs definitions.